### DISCUSSION TITLE
Making a Root Service Account

### DISCUSSION THREAD

**Discussion by rmuller-vu** [mia-platform/community#409] [Q&A]
### Description

We are trying to make a root service account. In the documentation ([here](https://docs.mia-platform.eu/docs/development_suite/identity-and-access-management/manage-service-accounts#creating-root-service-accounts)) it requires console-client-key and access token. Where do we get these? We created an account without passing console-client-key and access token. We got a 200 and a client id and secret successfully. We are trying to hit the apis at `/documentations/api-portal/` but are getting a 403 when using the token we generate using the service account we created. We don't see the root service account in the IAM for any of our companies and we can't add it to the company. Is there something we are missing? Is there a way through the UI to make root service accounts?

**Comment by davidebianchi** [comment]
Hello @rmuller-vu!
Root service accounts are not manageable via the UI, but only through the API. Additionally, root service accounts are not visible within a specific company because they are specific to the instance and are not shared across companies.
The client key should not be required anymore; I think the documentation needs to be updated. The access token must belong to a user with permission to create root service accounts (so with the console.root.serviceaccount.manage permission), or it must belong to another pre-existing root service account with this permission.
How did you create the service account without providing this information?

**Reply by rmuller-vu** [reply]
Where do we get our user access token? 
For the account we made. We used our cookie from the console browser.

**Reply by fredmaggiowski** [reply]
When you have an active session with the Console and use the API Portal, you are already using your access token (since the session cookie is shared among the domain) so there is no need to fiddle with cookies and whatnot.

If you wish to use a third party HTTP client that's fine, for that you'll have to retrieve your cookie from the browser (which as you noticed, holds your personal access token) since as of now the Console authentication flows do not support ROPG et similia. 

> [!TIP]
> The preferred way to use third party HTTP Client is to always use a service account but we understand it may be simpler to get your cookie and fire the api call.


Of course when speaking about creating root service account, at least the first one, must be created from a user with the `console.root.serviceaccount.manage` permission, as @davidebianchi said. In order to do so you will have to manually grant the permission to a user and then have it create the root service account (either via the API Portal or any third party HTTP client -postman,insomnia,bruno,curl...-), without doing so you'll receive the 403 error you mentioned in the original post.

The APIs you are interested in are the following:

<img width="1680" alt="image" src="https://github.com/mia-platform/community/assets/7142570/b34c28fe-4e5d-4f52-8c90-839b9a637306">

**Reply by zachhutchins** [reply]
@fredmaggiowski So I was able to use the endpoint, using my cookie was able to get a 200. Double checked and I have the permission `console.root.serviceaccount.manage`. When I attempt to use the root service account it has not access to any companies or projects. I gave it the following permissions.

For clarity I am using the [/m2m/token](https://docs.mia-platform.eu/docs/development_suite/identity-and-access-management/manage-service-accounts#service-account-authentication) endpoint to get the jwt 

```
[
"console.company.project.view",
"console.company.view",
"console.company.project.environment.view",
"console.company.providers.view",
"console.company.cluster.view"
,"marketplace.company.resources.view",
"console.project.view",
"console.project.environment.view",
"console.environment.view"
]
```

Those should be sufficient permissions to at the very least view all companies correct?

For example, when I make a call to `https://mia-console.vuhl.redchimney.com/api/user/companies` authenticated as the root account I get an empty array.

Do we have to give the root account access to each company manually? I notice the other service accounts that you can make for specific companies are available in the IAM, the root account does not show up.

**Reply by davidebianchi** [reply]
Hi @zachhutchins! The root service account should have those root permissions: https://docs.mia-platform.eu/docs/development_suite/identity-and-access-management/console-levels-and-permission-management#console-root-level-permissions

To view all companies, for example, should have `console.root.view` permission.
With the other permissions, you should give the access token manual access to all the resources (it is not possible by ui but only by backoffice, and it is not recommended).

**Reply by zachhutchins** [reply]
@davidebianchi So I went ahead and made a token with the following request body.

```
{
    "name":"root-sa",
    "permissions":
    [
        "console.root.view",
        "console.root.company.create",
        "console.root.company.delete",
        "console.root.project.create",
        "console.root.project.details.update",
        "console.root.project delete",
        "console.root.view",
        "console.root.user.bind",
        "console.root.user.manage",
        "console.root.serviceaccount.manage",
        "console.root.templates.manage",
        "console.root.features.manage",
        "marketplace.root.manage"
    ],
    "tokenEndpointAuthMethod":"client_secret_basic"
}
```

I got a 200 response.

```
{
    "clientId": "<ClientId>",
    "clientSecret": "<TOKEN>",
    "clientIdIssuedAt": 1709239429
}
```

When I use this sa I still get no companies, this sa should have almost no restrictions no?

For example, this curl returns no companies

```
curl --location 'https://local.mia.acme.com/api/user/companies' \
--header 'Authorization: Bearer <JWT TOKEN>'
```

**Reply by davidebianchi** [reply]
Oh yes, sorry for the confusion, you are right. At the moment, some root permissions only work on the backoffice side. The APIs that work with the root permissions are create/delete company, and the creation of the service account (so, `console.root.company.create`, `console.root.company.delete`, and `console.root.serviceaccount.manage` permissions).

When a service account creates a new company, it is set as the company owner by default. So, if a service account creates a company, it has all the permissions on the specific company.

You should assign the company owner role to specific companies through the back office by manually creating a binding associated with the company. Alternatively, you can create one service account per company so that it is manageable via the UI.

Does this align with your use case, or do you have some other specific needs?

**Reply by zachhutchins** [reply]
I think that answers my question, Thanks! :). There does not appear to be any endpoint to add a service account to a company, the only resource for adding users I see requires an email address. 

When you say manually do you mean via the backing database? We have existing companies the service account will need access to.

**Reply by fredmaggiowski** [reply]
You can find API documentation using the API Portal that is shipped with the Console, you can find it at `https://your-console/documentations/api-portal/` :)
